File: src/lib/api/grid.ts
import { invoke } from "./client";
import type { BrowseResult, GridCommitResult, GridEdit } from "./types";

// Typed wrappers over the grid commands — the ONLY invoke site for this domain.
export const gridApi = {
  browse: (sessionId: string, namespace: string, table: string, limit?: number) =>
    invoke<BrowseResult>("grid_browse", { sessionId, namespace, table, limit }),

  commit: (
    sessionId: string,
    namespace: string,
    table: string,
    keyColumns: string[],
    edits: GridEdit[],
  ) => invoke<GridCommitResult>("grid_commit", { sessionId, namespace, table, keyColumns, edits }),
};

File: src-tauri/src/commands/grid.rs
//! Thin grid command handlers: deserialize, call `grid_service`, map the error.

use tauri::State;

use crate::drivers::types::{BrowseResult, GridCommitResult, GridEdit};
use crate::services::grid_service;
use crate::state::AppState;
use crate::AppResult;

/// Loads the editable table-data view (a `limit + 1` window + row identity).
#[tauri::command]
pub async fn grid_browse(
    session_id: String,
    namespace: String,
    table: String,
    limit: Option<usize>,
    state: State<'_, AppState>,
) -> AppResult<BrowseResult> {
    grid_service::browse(&session_id, &namespace, &table, limit, &state.sessions).await
}

/// Commits a batch of staged edits in one transaction; any failure rolls back.
#[tauri::command]
pub async fn grid_commit(
    session_id: String,
    namespace: String,
    table: String,
    key_columns: Vec<String>,
    edits: Vec<GridEdit>,
    state: State<'_, AppState>,
) -> AppResult<GridCommitResult> {
    grid_service::commit(
        &session_id,
        &namespace,
        &table,
        key_columns,
        edits,
        &state.sessions,
    )
    .await
}
